README.md
# Endeavor Web Framework
### Status 

* Stable (v1.0.1)

## Overview 

Endeavor is a web framework written in Dart programming language and is inspired by frameworks like [Ruby on Rails](https://rubyonrails.org), and [Express Js](https://expressjs.com).

The goal of Endeavor framework is to provide developers effectively build backends in Dart. Currently, Dart is mainly used for developing client-side apps using Flutter. Our idea is to use Dart language for developing server-side applications also.  

## Quick Start 

### Prerequisites 

In order to use Endeavor you must have the [Dart SDK][dart_installation_link] installed on your machine.

This repository contains only core of Endeavor framework. There is no installation from this repo. So in order to use this framework , kindly use below instructions.

Endeavor requires Dart `">=2.17.0 <3.0.0"`


### Installing Endeavor CLI Tool 

```shell
#  Install the Endeavor cli from pub.dev

dart pub global activate ectl

```

### Creating a New Project 

Use the `ectl create <project name>` command to create a new project.

```shell
#  Create a new project called "MyDreamProject"
ectl create MyDreamProject
```

The directory structure is as follows:

```text
MyDreamProject
└── templates
    └── Home_view.dart
└── Controllers
    └── Home.dart
├── pubspec.yaml
├── main.dart
└── isolator.dart
```

### Start the Development Server

Next, open the newly directory of created project and start the dev server by:

```shell
#  Start the development server
ectl dev
```


By default port `80` is used. A custom port can be used inside the app's port option.

## Documentation

For detailed documentation, use our [documentation website](https://x-i-f-r-a.github.io/Endeavor-docs/)


## Versioning

Endeavor Framework is released by following the standard Semantic Versioning.


## Security Vulnerabilities or Bugs

If you discover a security vulnerability or bugs within Endeavor Framework, please send an e-mail to Gokul krishnan via gokulkrishnan5058@gmail.com. All security vulnerabilities and bugs will be promptly addressed.

## License

Endeavor Framework  is released under the BSD 3-Clause License.

[dart_installation_link]: https://dart.dev/get-dart


CHANGELOG.md
v0.1.8 (18-08-2022)

* Minor bug fixes

v0.1.9 (25-09-2022)

* Minor bug fixes

v1.0.0 (26-09-2022)

* Major bug patched
    - GET, POST, PUT not working together was the issue.
* Do not use older versions anymore. Use only from this latest version. 

v1.0.1 (21-10-2022)

* Minor bugs patched
    - Html form POST request giving base64 value for special characters.
    - Html post data is converted from key value to json data.
    - File uploading bugs is patched.
    
* Endeavor Framework is now stable from this version :tada: .
  - GET, POST, PUT not working together was the issue.
* Do not use older versions anymore. Use only from this latest version. 



test/test.dart
import 'package:endeavor/endeavor.dart';
void main() async {

              
  final app = Endeavor();


  // Html response

  app.GET('/', (Request req, Response res){
    res.HTML(

      '''
       <!doctype html>
        <html>
        <body>
          <form action="/download" method="post" enctype="multipart/form-data">
            <input type="file" name="file" /><br />
            
            <button type="submir">Save</button>
          </form>
        </body>
        </html>
      
      
      '''



    );
  });

  // Get POST data

  app.POST('/', (Request req, Response res) async {
    
    print(await req.body()); 

  });

  // Download a html form upload file

  app.POST('/download', (Request req, Response res) async {

    req.DownloadFile();

  });

  await app.runServer();

}



class RootController{

  home(Request req, Response res) async {

    res.sendString('T daa');
    
  }
}




example/test.dart

import 'dart:io';

import 'package:endeavor/endeavor.dart';

void main() async {

              
  final app = Endeavor();


  app.GET('/', (Request req, Response res){
    res.sendString('Ta daa');
  });

  app.GET('/home', RootController().home);

  
  var server = await HttpServer.bind('127.0.0.1', 80, shared: true);

  await app.runServer();

}



class RootController{


  home(Request req, Response res) async {

    res.sendString('T daa');
    
  }
}


lib/src/momentum.dart

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:mime/mime.dart';
import 'Http/request.dart';
import 'Http/response.dart';

class Momentum {

  Map? cors;
  Map? basicAuth;

  Momentum({this.cors=const{}, this.basicAuth=const{},});

  List<Function> functions = [];
  List<String> paths = [];


  Map routeKeys = new Map();
  Map methods = new Map();
  

  Future<HttpServer> createServer(int port, String host, {bool isShared = false}) async {
    
    return await HttpServer.bind(host, port, shared: isShared);

  }


  void createMap(){
    for(var i = 0; i < paths.length; i++){
    
      routeKeys[paths[i]] = functions[i];
    
    }
  }


  void _Handle(HttpRequest request){
    
    try{
      DateTime today = new DateTime.now();
      if(methods[request.uri.path.toString()].toString() == request.method.toString()){
       
       
        print('[ ${today.year.toString()}-${today.month.toString().padLeft(2,'0')}-${today.day.toString().padLeft(2,'0')} ${today.hour}:${today.minute}:${today.second.round()} ] ${methods[request.uri.path.toString()].toString()} ${request.uri.path.toString()}');
         
        
        final Response res = Response(request.response);
        final Request req = Request(request);


         if(basicAuth?.length==0){
            routeKeys[request.uri.path.toString()](req, res);
         }else{
          
           basicAuth?.forEach((key, value) async { 
            if(key.toString() == request.uri.path.toString()){
              final authHead = request.headers.value(HttpHeaders.authorizationHeader);
              if(authHead == null || authHead == ''){
                request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                request.response.statusCode = 401;
                request.response.write('401 Unauthorized');
                await request.response.close();
              }else
              {
                final data = utf8.decode(base64.decode(authHead.replaceAll('Basic ', ''))).toString();

                if(data.toString()==value.toString()){
                  routeKeys[request.uri.path.toString()](req, res);
                }else{
                  request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                  request.response.statusCode = 401;
                  request.response.write('401 Unauthorized');
                  await request.response.close();
                }
              
              }
            }else{
              
              routeKeys[request.uri.path.toString()](req, res);
              
            }
          });
         }
        
      }else{
        request.response.statusCode = HttpStatus.methodNotAllowed;
        request.response.write('Unsupported HTTP Method');
        request.response.close();
      }
      
      

    }catch(e){
      CustomException('[ Error ] ${e.toString()}').showError();
      
    }
  }


  Future<void> _HandleHttpRequest(HttpServer server, bool compress) async {
    server.autoCompress = compress;
    server.serverHeader = 'Momentum-V0.0.1';

    await for(HttpRequest request in server)  {
      
      
      try{
        
        if( paths.contains(request.uri.path.toString()) || request.uri.hasQuery){
         

          if(cors?.length!=0){
            cors?.forEach((key, value) { 
              request.response.headers.set(key, value);
            });
            
            
            this.createMap();
            _Handle(request);
          }else{
            
            this.createMap();
            _Handle(request);
          }
    
        }else{
          _serveStatic(request);
        }
       

    } on HttpException catch(e) {
      throw e;
    }

      
    }
  }


  Future GET(String path, Function fn) async {
    paths.add(path);
    functions.add(fn);
    methods[path] = 'GET';
    
  }

  Future POST(String path, Function fn) async {
    paths.add(path);
    functions.add(fn);
    methods[path] = 'POST';
    
  }

  Future PUT(String path, Function fn) async {
    paths.add(path);
    functions.add(fn);
    methods[path] = 'PUT';
    
  }

  Future PATCH(String path, Function fn) async {
    paths.add(path);
    functions.add(fn);
    methods[path] = 'PATCH';
    
  }

  Future UPDATE(String path, Function fn) async {
    paths.add(path);
    functions.add(fn);
    methods[path] = 'UPDATE';
    
  }

  Future DELETE(String path, Function fn) async {
    paths.add(path);
    functions.add(fn);
    methods[path] = 'DELETE';
    
  }

  Future _serveStatic(HttpRequest req) async {
    String _rootPath = Directory.current.path;
      File staticFile = File(_rootPath+'/templates/'+req.uri.path);
      await staticFile.exists().then((bool found) async {
        if (found) {
          req.response.statusCode = HttpStatus.ok;
          var mimeType = lookupMimeType(_rootPath+'/templates/'+req.uri.path);
          req.response.headers.set('Content-Type', mimeType!);
          staticFile.openRead().pipe(req.response).catchError((e) { 
            throw CustomException(e.toString()).showError();
          });
        } else {
          req.response.statusCode = HttpStatus.notFound;
          req.response.headers.contentType = ContentType.html;
          req.response.write('<br><center><h2>Not Found</h2></center>');
          await req.response.close();
        }
      });
  }
  
   Future<void> runServer( {final server='', String? host='127.0.0.1', int port=80, bool compress = true, bool isShared = false}) async {

    
    
    if(server is HttpServer){
      
      print('[ INFO ] SERVER STARTED');
      
      _HandleHttpRequest(server, compress);
      
    
    }else{
      if(host==''){
        final address = InternetAddress.loopbackIPv4;
        final createserver = await createServer(port, address.toString(), isShared: isShared );
        print('[ INFO ] SERVER STARTED AT ${address.toString()}:$port/');
        
        _HandleHttpRequest(createserver, compress);
        
      }else{
        final createserver = await createServer(port, host.toString(), isShared: isShared );
        print('[ INFO ] SERVER STARTED AT http://$host:$port/');
        
        _HandleHttpRequest(createserver, compress);

      }
    }
    
    
  }




}





class CustomException implements Exception {
  String? reason;
  CustomException(this.reason);

  showError(){
   return reason!;
  }
}




lib/src/Database/momentum_mysql.dart
import 'dart:async';

import 'package:mysql1/mysql1.dart';

class MomentumMySql{

  String? host;
  int? port;
  String? username;
  String? password;
  String? databaseName;
  MySqlConnection? _con;

  MomentumMySql({required this.host,required this.port,required this.username,required this.password,required this.databaseName}){
    
   connectDB();

  }

  void connectDB() async {
     var settings = new ConnectionSettings(
      host: host!, 
      port: port!,
      user: username!,
      password: password,
      db: databaseName
    );
    try{
      var conn = await MySqlConnection.connect(settings);
      print('[Connected to MySql server]');
      _con=conn;
    } on MySqlException catch(e){
      throw e;
    }
  }

  FutureOr InsertAll(String QUERY, List DATA) async {
    return await _con?.query(QUERY, DATA);
  }

  FutureOr FetchAll(String TABLE) async {
    final data = await _con?.query('SELECT * FROM $TABLE;');
    return data;
  }

  FutureOr FetchByCondition(String QUERY) async {
    final data = await _con?.query('$QUERY;');
    return data;
  }


  FutureOr Update(String QUERY, List DATA) async {
    await _con?.query(QUERY, DATA);
  }


}

lib/src/endeavor.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:mime/mime.dart';
import 'Http/request.dart';
import 'Http/response.dart';

class Endeavor {

  Map? cors;
  Map? basicAuth;

  Endeavor({this.cors=const{}, this.basicAuth=const{},});

  List<Function> _get_functions = [];
  List<Function> _post_functions = [];
  List<Function> _put_functions = [];
  List<Function> _update_functions = [];
  List<Function> _delete_functions = [];
  List<Function> _patch_functions = [];


  List<String> _getpaths = [];
  List<String> _postpaths = [];
  List<String> _putpaths = [];
  List<String> _updatepaths = [];
  List<String> _deletepaths = [];
  List<String> _patchpaths = [];




  Map _getrouteKeys = new Map();
  Map _postrouteKeys = new Map();
  Map _putrouteKeys = new Map();
  Map _updaterouteKeys = new Map();
  Map _deleterouteKeys = new Map();
  Map _patchrouteKeys = new Map();
  Map _getmethod = new Map();
  Map _postmethod = new Map();
  Map _putmethod = new Map();
  Map _deletemethod = new Map();
  Map _updatemethod = new Map();
  Map _patchmethod = new Map();
  

  Future<HttpServer> createServer(int port, String host, {bool isShared = false}) async {
    
    return await HttpServer.bind(host, port, shared: isShared);

  }


  void createGETMap(){
    for(var i = 0; i < _getpaths.length; i++){
    
      _getrouteKeys[_getpaths[i]] = _get_functions[i];
    
    }
  }

  void createPOSTMap(){
    for(var i = 0; i < _postpaths.length; i++){
    
      _postrouteKeys[_postpaths[i]] = _post_functions[i];
    
    }
  }

  void createPUTMap(){
    for(var i = 0; i < _putpaths.length; i++){
    
      _putrouteKeys[_putpaths[i]] = _put_functions[i];
    
    }
  }

  void createUPDATEMap(){
    for(var i = 0; i < _updatepaths.length; i++){
    
      _updaterouteKeys[_updatepaths[i]] = _update_functions[i];
    
    }
  }

  void createDELETEMap(){
    for(var i = 0; i < _deletepaths.length; i++){
    
      _deleterouteKeys[_deletepaths[i]] = _delete_functions[i];
    
    }

  }

  void createPATCHMap(){
    for(var i = 0; i < _patchpaths.length; i++){
    
      _patchrouteKeys[_patchpaths[i]] = _patch_functions[i];
    
    }

  }



  void _handleGET(HttpRequest request, String route) async {

    try{
      DateTime today = new DateTime.now();
      if(
        _getmethod[request.uri.path.toString()].toString() == request.method.toString()
        
      ){
       
       
        print('[ ${today.year.toString()}-${today.month.toString().padLeft(2,'0')}-${today.day.toString().padLeft(2,'0')} ${today.hour}:${today.minute}:${today.second.round()} ] ${ request.method.toString()} ${request.uri.path.toString()}');
         
        
        final Response res = Response(request.response);
        final Request req = Request(request);


         if(basicAuth?.length==0){
            _getrouteKeys[request.uri.path.toString()](req, res);
         }else{
          
           basicAuth?.forEach((key, value) async { 
            if(key.toString() == request.uri.path.toString()){
              final authHead = request.headers.value(HttpHeaders.authorizationHeader);
              if(authHead == null || authHead == ''){
                request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                request.response.statusCode = 401;
                request.response.write('401 Unauthorized');
                await request.response.close();
              }else
              {
                final data = utf8.decode(base64.decode(authHead.replaceAll('Basic ', ''))).toString();

                if(data.toString()==value.toString()){
                  _getrouteKeys[request.uri.path.toString()](req, res);
                }else{
                  request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                  request.response.statusCode = 401;
                  request.response.write('401 Unauthorized');
                  await request.response.close();
                }
              
              }
            }else{
              
              _getrouteKeys[request.uri.path.toString()](req, res);
              
            }
          });
         }
        
      }else{
        request.response.statusCode = HttpStatus.methodNotAllowed;
        request.response.write('Unsupported HTTP Method');
        request.response.close();
      }
      
      

    }catch(e){
      CustomException('[ Error ] ${e.toString()}').showError();
      
    }

  }

  void _handlePOST(HttpRequest request, String route) async {

    try{
      DateTime today = new DateTime.now();
      if(
        _postmethod[request.uri.path.toString()].toString() == request.method.toString()
        
      ){
       
       
        print('[ ${today.year.toString()}-${today.month.toString().padLeft(2,'0')}-${today.day.toString().padLeft(2,'0')} ${today.hour}:${today.minute}:${today.second.round()} ] ${ request.method.toString()} ${request.uri.path.toString()}');
         
        
        final Response res = Response(request.response);
        final Request req = Request(request);


         if(basicAuth?.length==0){
            _postrouteKeys[request.uri.path.toString()](req, res);
         }else{
          
           basicAuth?.forEach((key, value) async { 
            if(key.toString() == request.uri.path.toString()){
              final authHead = request.headers.value(HttpHeaders.authorizationHeader);
              if(authHead == null || authHead == ''){
                request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                request.response.statusCode = 401;
                request.response.write('401 Unauthorized');
                await request.response.close();
              }else
              {
                final data = utf8.decode(base64.decode(authHead.replaceAll('Basic ', ''))).toString();

                if(data.toString()==value.toString()){
                  _postrouteKeys[request.uri.path.toString()](req, res);
                }else{
                  request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                  request.response.statusCode = 401;
                  request.response.write('401 Unauthorized');
                  await request.response.close();
                }
              
              }
            }else{
              
              _postrouteKeys[request.uri.path.toString()](req, res);
              
            }
          });
         }
        
      }else{
        request.response.statusCode = HttpStatus.methodNotAllowed;
        request.response.write('Unsupported HTTP Method');
        request.response.close();
      }
      
      

    }catch(e){
      CustomException('[ Error ] ${e.toString()}').showError();
      
    }

  }

  void _handleDELETE(HttpRequest request, String route) async {

    try{
      DateTime today = new DateTime.now();
      if(
        _deletemethod[request.uri.path.toString()].toString() == request.method.toString()
        
      ){
       
       
        print('[ ${today.year.toString()}-${today.month.toString().padLeft(2,'0')}-${today.day.toString().padLeft(2,'0')} ${today.hour}:${today.minute}:${today.second.round()} ] ${ request.method.toString()} ${request.uri.path.toString()}');
         
        
        final Response res = Response(request.response);
        final Request req = Request(request);


         if(basicAuth?.length==0){
            _deleterouteKeys[request.uri.path.toString()](req, res);
         }else{
          
           basicAuth?.forEach((key, value) async { 
            if(key.toString() == request.uri.path.toString()){
              final authHead = request.headers.value(HttpHeaders.authorizationHeader);
              if(authHead == null || authHead == ''){
                request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                request.response.statusCode = 401;
                request.response.write('401 Unauthorized');
                await request.response.close();
              }else
              {
                final data = utf8.decode(base64.decode(authHead.replaceAll('Basic ', ''))).toString();

                if(data.toString()==value.toString()){
                  _deleterouteKeys[request.uri.path.toString()](req, res);
                }else{
                  request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                  request.response.statusCode = 401;
                  request.response.write('401 Unauthorized');
                  await request.response.close();
                }
              
              }
            }else{
              
              _deleterouteKeys[request.uri.path.toString()](req, res);
              
            }
          });
         }
        
      }else{
        request.response.statusCode = HttpStatus.methodNotAllowed;
        request.response.write('Unsupported HTTP Method');
        request.response.close();
      }
      
      

    }catch(e){
      CustomException('[ Error ] ${e.toString()}').showError();
      
    }

  }

  void _handlePUT(HttpRequest request, String route) async {

    try{
      DateTime today = new DateTime.now();
      if(
        _putmethod[request.uri.path.toString()].toString() == request.method.toString()
        
      ){
       
       
        print('[ ${today.year.toString()}-${today.month.toString().padLeft(2,'0')}-${today.day.toString().padLeft(2,'0')} ${today.hour}:${today.minute}:${today.second.round()} ] ${ request.method.toString()} ${request.uri.path.toString()}');
         
        
        final Response res = Response(request.response);
        final Request req = Request(request);


         if(basicAuth?.length==0){
            _putrouteKeys[request.uri.path.toString()](req, res);
         }else{
          
           basicAuth?.forEach((key, value) async { 
            if(key.toString() == request.uri.path.toString()){
              final authHead = request.headers.value(HttpHeaders.authorizationHeader);
              if(authHead == null || authHead == ''){
                request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                request.response.statusCode = 401;
                request.response.write('401 Unauthorized');
                await request.response.close();
              }else
              {
                final data = utf8.decode(base64.decode(authHead.replaceAll('Basic ', ''))).toString();

                if(data.toString()==value.toString()){
                  _putrouteKeys[request.uri.path.toString()](req, res);
                }else{
                  request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                  request.response.statusCode = 401;
                  request.response.write('401 Unauthorized');
                  await request.response.close();
                }
              
              }
            }else{
              
              _putrouteKeys[request.uri.path.toString()](req, res);
              
            }
          });
         }
        
      }else{
        request.response.statusCode = HttpStatus.methodNotAllowed;
        request.response.write('Unsupported HTTP Method');
        request.response.close();
      }
      
      

    }catch(e){
      CustomException('[ Error ] ${e.toString()}').showError();
      
    }

  }

  void _handleUPDATE(HttpRequest request, String route) async {

    try{
      DateTime today = new DateTime.now();
      if(
        _updatemethod[request.uri.path.toString()].toString() == request.method.toString()
        
      ){
       
       
        print('[ ${today.year.toString()}-${today.month.toString().padLeft(2,'0')}-${today.day.toString().padLeft(2,'0')} ${today.hour}:${today.minute}:${today.second.round()} ] ${ request.method.toString()} ${request.uri.path.toString()}');
         
        
        final Response res = Response(request.response);
        final Request req = Request(request);


         if(basicAuth?.length==0){
            _updaterouteKeys[request.uri.path.toString()](req, res);
         }else{
          
           basicAuth?.forEach((key, value) async { 
            if(key.toString() == request.uri.path.toString()){
              final authHead = request.headers.value(HttpHeaders.authorizationHeader);
              if(authHead == null || authHead == ''){
                request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                request.response.statusCode = 401;
                request.response.write('401 Unauthorized');
                await request.response.close();
              }else
              {
                final data = utf8.decode(base64.decode(authHead.replaceAll('Basic ', ''))).toString();

                if(data.toString()==value.toString()){
                  _updaterouteKeys[request.uri.path.toString()](req, res);
                }else{
                  request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                  request.response.statusCode = 401;
                  request.response.write('401 Unauthorized');
                  await request.response.close();
                }
              
              }
            }else{
              
              _updaterouteKeys[request.uri.path.toString()](req, res);
              
            }
          });
         }
        
      }else{
        request.response.statusCode = HttpStatus.methodNotAllowed;
        request.response.write('Unsupported HTTP Method');
        request.response.close();
      }
      
      

    }catch(e){
      CustomException('[ Error ] ${e.toString()}').showError();
      
    }

  }

  void _handlePATCH(HttpRequest request, String route) async {

    try{
      DateTime today = new DateTime.now();
      if(
        _patchmethod[request.uri.path.toString()].toString() == request.method.toString()
        
      ){
       
       
        print('[ ${today.year.toString()}-${today.month.toString().padLeft(2,'0')}-${today.day.toString().padLeft(2,'0')} ${today.hour}:${today.minute}:${today.second.round()} ] ${ request.method.toString()} ${request.uri.path.toString()}');
         
        
        final Response res = Response(request.response);
        final Request req = Request(request);


         if(basicAuth?.length==0){
            _patchrouteKeys[request.uri.path.toString()](req, res);
         }else{
          
           basicAuth?.forEach((key, value) async { 
            if(key.toString() == request.uri.path.toString()){
              final authHead = request.headers.value(HttpHeaders.authorizationHeader);
              if(authHead == null || authHead == ''){
                request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                request.response.statusCode = 401;
                request.response.write('401 Unauthorized');
                await request.response.close();
              }else
              {
                final data = utf8.decode(base64.decode(authHead.replaceAll('Basic ', ''))).toString();

                if(data.toString()==value.toString()){
                  _patchrouteKeys[request.uri.path.toString()](req, res);
                }else{
                  request.response.headers.set('WWW-Authenticate', 'Basic realm="Secure Area"');
                  request.response.statusCode = 401;
                  request.response.write('401 Unauthorized');
                  await request.response.close();
                }
              
              }
            }else{
              
              _patchrouteKeys[request.uri.path.toString()](req, res);
              
            }
          });
         }
        
      }else{
        request.response.statusCode = HttpStatus.methodNotAllowed;
        request.response.write('Unsupported HTTP Method');
        request.response.close();
      }
      
      

    }catch(e){
      CustomException('[ Error ] ${e.toString()}').showError();
      
    }
  }




  void _Handle(HttpRequest request){
  
    
    switch(request.method.toString()){


      case 'GET':
        _handleGET(request, request.uri.path.toString());
        break;

      case 'POST':
        _handlePOST(request, request.uri.path.toString());
        break;

      case 'PUT':
        _handlePUT(request, request.uri.path.toString());
        break;

      case 'UPDATE':
        _handleUPDATE(request, request.uri.path.toString());
        break;

      case 'DELETE':
        _handleDELETE(request, request.uri.path.toString());
        break;

      case 'PATCH':
        _handlePATCH(request, request.uri.path.toString());
        break;

      default:
        request.response.write('Unsupported HTTP Method');
        request.response.close();
        break;
        
    }
    

    
  }


  Future<void> _HandleHttpRequest(HttpServer server, bool compress) async {
    server.autoCompress = compress;
    server.serverHeader = 'Endeavor';

    await for(HttpRequest request in server)  {
      
      try{
        
        switch(request.method.toString()){

          case 'GET': {
            if( _getpaths.contains(request.uri.path.toString()) || request.uri.hasQuery){
         

              if(cors?.length!=0){
                cors?.forEach((key, value) { 
                  request.response.headers.set(key, value);
                });
                
                
                this.createGETMap();
                _Handle(request);
              }else{
                
                this.createGETMap();
                _Handle(request);
              }
    
            }else{
              _serveStatic(request);
            }
          }
            
          break;

          case 'POST': {
            if( _postpaths.contains(request.uri.path.toString()) || request.uri.hasQuery){
         

              if(cors?.length!=0){
                cors?.forEach((key, value) { 
                  request.response.headers.set(key, value);
                });
                
                
                this.createPOSTMap();
                _Handle(request);
              }else{
                
                this.createPOSTMap();
                _Handle(request);
              }
    
            }else{
              _serveStatic(request);
            }
          }
          break;
          case 'PUT': {
            if( _putpaths.contains(request.uri.path.toString()) || request.uri.hasQuery){
         

              if(cors?.length!=0){
                cors?.forEach((key, value) { 
                  request.response.headers.set(key, value);
                });
                
                
                this.createPUTMap();
                _Handle(request);
              }else{
                
                this.createPUTMap();
                _Handle(request);
              }
    
            }else{
              _serveStatic(request);
            }
          }
          break;
          case 'UPDATE': {
            if( _updatepaths.contains(request.uri.path.toString()) || request.uri.hasQuery){
         

              if(cors?.length!=0){
                cors?.forEach((key, value) { 
                  request.response.headers.set(key, value);
                });
                
                
                this.createUPDATEMap();
                _Handle(request);
              }else{
                
                this.createUPDATEMap();
                _Handle(request);
              }
    
            }else{
              _serveStatic(request);
            }
          }
          break;
          case 'DELETE': {
            if( _deletepaths.contains(request.uri.path.toString()) || request.uri.hasQuery){
         

              if(cors?.length!=0){
                cors?.forEach((key, value) { 
                  request.response.headers.set(key, value);
                });
                
                
                this.createDELETEMap();
                _Handle(request);
              }else{
                
                this.createDELETEMap();
                _Handle(request);
              }
    
            }else{
              _serveStatic(request);
            }
          }
          break;

        case 'PATCH': {
            if( _patchpaths.contains(request.uri.path.toString()) || request.uri.hasQuery){
         

              if(cors?.length!=0){
                cors?.forEach((key, value) { 
                  request.response.headers.set(key, value);
                });
                
                
                this.createPATCHMap();
                _Handle(request);
              }else{
                
                this.createPATCHMap();
                _Handle(request);
              }
    
            }else{
              _serveStatic(request);
            }
          }
          break;

          default:
            break;
          
        }
       

    } on HttpException catch(e) {
      throw e;
    }

      
    }
  }


  Future GET(String path, Function fn) async {
    _getpaths.add(path);
    _get_functions.add(fn);
    _getmethod[path] = 'GET';
    
  }

  Future POST(String path, Function fn) async {
    _postpaths.add(path);
    _post_functions.add(fn);
    _postmethod[path] = 'POST';
    
  }

  Future PUT(String path, Function fn) async {
    _putpaths.add(path);
    _put_functions.add(fn);
    _putmethod[path] = 'PUT';
    
  }

  Future PATCH(String path, Function fn) async {
    _patchpaths.add(path);
    _patch_functions.add(fn);
    _patchmethod[path] = 'PATCH';
    
  }

  Future UPDATE(String path, Function fn) async {
    _updatepaths.add(path);
    _update_functions.add(fn);
    _updatemethod[path] = 'UPDATE';
    
  }

  Future DELETE(String path, Function fn) async {
    _deletepaths.add(path);
    _delete_functions.add(fn);
    _deletemethod[path] = 'DELETE';
    
  }

  Future _serveStatic(HttpRequest req) async {
    String _rootPath = Directory.current.path;
      File staticFile = File(_rootPath+'/templates/'+req.uri.path);
      await staticFile.exists().then((bool found) async {
        if (found) {
          req.response.statusCode = HttpStatus.ok;
          var mimeType = lookupMimeType(_rootPath+'/templates/'+req.uri.path);
          req.response.headers.set('Content-Type', mimeType!);
          staticFile.openRead().pipe(req.response).catchError((e) { 
            throw CustomException(e.toString()).showError();
          });
        } else {
          req.response.statusCode = HttpStatus.notFound;
          req.response.headers.contentType = ContentType.html;
          req.response.write('<br><center><h2>Not Found</h2></center>');
          await req.response.close();
        }
      });
  }
  
   Future<void> runServer( {final server='', String? host='127.0.0.1', int port=80, bool compress = true, bool isShared = false}) async {

    
    
    if(server is HttpServer){
      
      print('[ INFO ] SERVER STARTED');
      
      _HandleHttpRequest(server, compress);
      
    
    }else{
      if(host==''){
        final address = InternetAddress.loopbackIPv4;
        final createserver = await createServer(port, address.toString(), isShared: isShared );
        print('[ INFO ] SERVER STARTED AT ${address.toString()}:$port/');
        
        _HandleHttpRequest(createserver, compress);
        
      }else{
        final createserver = await createServer(port, host.toString(), isShared: isShared );
        print('[ INFO ] SERVER STARTED AT http://$host:$port/');
        
        _HandleHttpRequest(createserver, compress);

      }
    }
    
    
  }




}


class CustomException implements Exception {
  String? reason;
  CustomException(this.reason);

  showError(){
   return reason!;
  }
}


